### ISSUE TITLE
Adding new markup control points is very slow if control points table is visible

### ISSUE THREAD

**Issue by lassoan** [Slicer/Slicer#6379]
Adding new markup control points is very slow if Markups module is active and control points table is visible.

See user error report at https://discourse.slicer.org/t/adding-points-to-point-list-markup-becomes-progressively-slower/23419/4

## Steps to reproduce

- Create a `Point list` markup
- Add 100 control points
- Add one more control point - it takes a fraction of a second
- Go to Markups module, open Control points section
- Add one more control point - it takes several seconds

## Expected behavior

Adding a control point should only take a fraction of a second.

## Environment
- Slicer version: Slicer-5.1.0-2022-05-08
- Operating system: Windows10

**Comment by lassoan** [comment]
@smrolfe could you have a look at this? Update of the markups control points table is so slow when you have more than a few ten points that it becomes very hard to use it.

**Comment by smrolfe** [comment]
Thanks, I will look into this.

**Comment by jamesobutler** [comment]
Every row in the control points table is getting called to be updated when updateWidgetFromMRML is being processed. This is increasing the amount of processing required when the number of control points in the table gets larger.

https://github.com/Slicer/Slicer/blob/dcd132b264a741e01ae42ce35d774061bb8fc6c5/Modules/Loadable/Markups/qSlicerMarkupsModuleWidget.cxx#L1075-L1085

**Comment by lassoan** [comment]
I'm experimenting with a model-view-based markup control point table (similar to segment list table). It should be much faster as the current one, but I will know more by tomorrow.

@smrolfe if you can figure out a quick and simple fix for the current code then it's good, but if it looks difficult then you may consider waiting for the result of my experiment before investing a lot of time in it.

**Comment by smrolfe** [comment]
Thanks @lassoan sounds like a good plan. I'll update this thread if I find a quick fix this afternoon.

**Comment by jcfr** [comment]
> model based table

That seems like a sensible path forward.


Once this is addressed, another good candidate for v5.0.3 patch release I think

**Comment by lassoan** [comment]
I've implemented a [prototype for a control points table view](https://github.com/lassoan/Slicer/tree/attic/markups-control-points-table-view) based on how the segments table is implemented. It is nice that it makes search&filtering simpler and allows reuse of the table in other places, but it still creates all model items for each control point, which would make the table still too slow if we have many thousand control points. It would be better to if we implemented a special table model instead, similarly to pqSpreadSheetViewModel, which would only create items for the items requested by the view.

This all would be a lot of work, and for a few thousand points we can make the current widget work with some simple optimizations. I'll push a fix with the proposed changes today.
